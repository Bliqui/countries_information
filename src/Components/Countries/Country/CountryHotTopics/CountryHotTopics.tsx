import {
  Box,
  Center,
  Flex,
  Heading,
  Image,
  Link,
  SimpleGrid,
  Spinner,
  Text,
  VStack,
} from "@chakra-ui/react";
import { useFetch } from "../../../../hooks/useFetch";
import { formateDate } from "../../../../utils/formateDate";
import { textLengthFormate } from "../../../../utils/textLengthFormate";

type TopicsData = {
  articles: [
    {
      source: {
        name: string;
      };
      author: string;
      title: string;
      description: string;
      url: string;
      urlToImage: string;
      publishedAt: string;
      countryFlag: string;
    }
  ];
};

export const CountryHotTopics = ({
  countryFlag,
  countryIso,
}: {
  countryFlag: string;
  countryIso: string;
}) => {
  const {
    data: receivedTopics,
    isLoading,
    error,
  } = useFetch<TopicsData>(
    `https://newsapi.org/v2/top-headlines?country=${countryIso}&apiKey=${process.env.REACT_APP_API_KEY}`
  );

  const cutData = receivedTopics?.articles.slice(0, 6);

  return cutData?.length ? (
    <Center mt="40px" flexDir="column">
      <Heading>Top-6 country news:</Heading>
      <SimpleGrid
        mt="40px"
        width="100%"
        gridTemplateColumns="repeat(auto-fit, minmax(300px , 1fr));"
        spacing="30px"
      >
        {isLoading && <Spinner />}
        {receivedTopics?.articles ? (
          cutData?.map((article) => {
            return (
              <Flex
                key={article.title}
                _dark={{ bgColor: "gray.800" }}
                borderRadius="10px"
                alignItems="center"
                w="320px"
                minH="400px"
                mx="auto"
                boxShadow="0px 0px 8px 0px rgba(66, 68, 90, .6)"
                flexDir="column"
                textAlign="center"
              >
                <Image
                  src={article.urlToImage ? article.urlToImage : countryFlag}
                  borderTopLeftRadius="10px"
                  borderTopRightRadius="10px"
                />
                <VStack p="10px" h="100%" justifyContent="space-between">
                  <Heading mb="10px" size="md">
                    {textLengthFormate(article.title, 70)}
                  </Heading>
                  <Text>{textLengthFormate(article.description, 120)}</Text>
                  <Box>
                    <Text mt="10px" mb="5px" fontWeight={600}>
                      {article.author && `Author: ${article.author}`}
                    </Text>
                    <Text textTransform="capitalize" mb="10px" fontWeight={600}>
                      {formateDate(article.publishedAt)}
                    </Text>
                    <Link
                      py="8px"
                      px="20px"
                      mt="10px"
                      borderRadius="10px"
                      _hover={{
                        textTransform: "none",
                        bgColor: "whiteAlpha.900",
                      }}
                      _active={{ bgColor: "blackAlpha.400" }}
                      _dark={{
                        bgColor: "blackAlpha.500",
                        _hover: { bgColor: "whiteAlpha.200" },
                      }}
                      boxShadow="0px 0px 4px 0px rgba(66, 68, 90, 1)"
                      bgColor="whiteAlpha.700"
                      isExternal
                      href={article.url}
                      display="inline-block"
                    >
                      Read full article
                    </Link>
                  </Box>
                </VStack>
              </Flex>
            );
          })
        ) : (
          <Text>{error}</Text>
        )}
      </SimpleGrid>
    </Center>
  ) : null;
};
