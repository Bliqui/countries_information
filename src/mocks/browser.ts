import { rest } from "msw";
import { setupServer } from "msw/node";

type AllCountriesDataType = {
  name: string;
  population: number;
  flag: string;
  region: string;
  capital: string;
};

type CountryPageDataType = {
  country: {
    name: string;
    capital: string;
    nativeName: string;
    population: number;
    region: string;
    subRegion: string;
    topLevelDomain: string[];
    currencies: { code: string; name: string; symbol: string }[];
    languages: {
      name: string;
      nativeName: string;
    }[];
    flag: string;
    borders: string[];
    latlng: [number, number];
  };
};

export const server = setupServer(
  rest.get<AllCountriesDataType>(
    "https://restcountries.com/v3.1/all",
    (req, res, ctx) => {
      return res(
        ctx.json([
          {
            name: { common: "Poland" },
            population: 40000000,
            flags: { png: "url" },
            region: "Europe",
            capital: ["Warsaw"],
          },
          {
            name: { common: "Russia" },
            population: 60000000,
            flags: { png: "url" },
            region: "Europe",
            capital: ["Moscow"],
          },
        ])
      );
    }
  ),
  rest.get<AllCountriesDataType>(
    `https://newsapi.org/v2/top-headlines?country=RUS`,
    (req, res, ctx) => {
      return res(
        ctx.json({
          articles: [
            {
              author: "Petya",
              title: "Shock content",
              description: "Shock content happened",
              url: "some url",
              publishedAt: "2022-08-29T08:00:28Z",
            },
            {
              author: "Vasya",
              title: "extra Shock content",
              description: "Shock content happened twice",
              url: "some url2",
              publishedAt: "2022-09-29T08:00:28Z",
            },
          ],
        })
      );
    }
  )
);
// Describe the requests to mock.
