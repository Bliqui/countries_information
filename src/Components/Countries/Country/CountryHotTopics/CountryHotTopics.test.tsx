import { screen, render } from "@testing-library/react";
import { server } from "../../../../mocks/browser";
import { CountryHotTopics } from "./CountryHotTopics";

beforeAll(() => {
  server.listen();
});

afterEach(() => {
  server.resetHandlers();
});

afterAll(() => {
  server.close();
});

describe("hot topics component", () => {
  it("renders properly", async () => {
    render(<CountryHotTopics countryFlag={"flag"} countryIso={"alpha2Code"} />);

    expect(await screen.findByText("Author: Petya")).toBeInTheDocument();
    expect(await screen.findByText("Shock content")).toBeInTheDocument();
    expect(
      await screen.findByText("Shock content happened")
    ).toBeInTheDocument();
    expect(
      await screen.findByText("Monday, August 29, 2022 at 10:00:28 AM GMT+2")
    ).toBeInTheDocument();

    expect(await screen.findByText("Author: Vasya")).toBeInTheDocument();
    expect(await screen.findByText("extra Shock content")).toBeInTheDocument();
    expect(
      await screen.findByText("Shock content happened twice")
    ).toBeInTheDocument();
    expect(
      await screen.findByText(
        "Thursday, September 29, 2022 at 10:00:28 AM GMT+2"
      )
    ).toBeInTheDocument();
  });

  it("matches snapshot", async () => {
    const output = render(
      <CountryHotTopics countryFlag={"flag"} countryIso={"alpha2Code"} />
    );
    expect(await screen.findByText("Author: Vasya")).toBeInTheDocument();
    expect(output).toMatchSnapshot();
  });
});
