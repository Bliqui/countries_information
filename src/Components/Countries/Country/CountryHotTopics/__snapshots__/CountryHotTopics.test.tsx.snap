// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hot topics component matches snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="css-0"
      >
        <h2
          class="chakra-heading css-0"
        >
          Top-6 country news:
        </h2>
        <div
          class="css-0"
        >
          <div
            class="css-0"
          >
            <img
              class="chakra-image css-0"
              src="flag"
            />
            <div
              class="chakra-stack css-0"
            >
              <h2
                class="chakra-heading css-0"
              >
                Shock content
              </h2>
              <p
                class="chakra-text css-0"
              >
                Shock content happened
              </p>
              <div
                class="css-0"
              >
                <p
                  class="chakra-text css-0"
                >
                  Author: Petya
                </p>
                <p
                  class="chakra-text css-0"
                >
                  Monday, August 29, 2022 at 10:00:28 AM GMT+2
                </p>
                <a
                  class="chakra-link css-0"
                  href="some url"
                  rel="noopener"
                  target="_blank"
                >
                  Read full article
                </a>
              </div>
            </div>
          </div>
          <div
            class="css-0"
          >
            <img
              class="chakra-image css-0"
              src="flag"
            />
            <div
              class="chakra-stack css-0"
            >
              <h2
                class="chakra-heading css-0"
              >
                extra Shock content
              </h2>
              <p
                class="chakra-text css-0"
              >
                Shock content happened twice
              </p>
              <div
                class="css-0"
              >
                <p
                  class="chakra-text css-0"
                >
                  Author: Vasya
                </p>
                <p
                  class="chakra-text css-0"
                >
                  Thursday, September 29, 2022 at 10:00:28 AM GMT+2
                </p>
                <a
                  class="chakra-link css-0"
                  href="some url2"
                  rel="noopener"
                  target="_blank"
                >
                  Read full article
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="css-0"
    >
      <h2
        class="chakra-heading css-0"
      >
        Top-6 country news:
      </h2>
      <div
        class="css-0"
      >
        <div
          class="css-0"
        >
          <img
            class="chakra-image css-0"
            src="flag"
          />
          <div
            class="chakra-stack css-0"
          >
            <h2
              class="chakra-heading css-0"
            >
              Shock content
            </h2>
            <p
              class="chakra-text css-0"
            >
              Shock content happened
            </p>
            <div
              class="css-0"
            >
              <p
                class="chakra-text css-0"
              >
                Author: Petya
              </p>
              <p
                class="chakra-text css-0"
              >
                Monday, August 29, 2022 at 10:00:28 AM GMT+2
              </p>
              <a
                class="chakra-link css-0"
                href="some url"
                rel="noopener"
                target="_blank"
              >
                Read full article
              </a>
            </div>
          </div>
        </div>
        <div
          class="css-0"
        >
          <img
            class="chakra-image css-0"
            src="flag"
          />
          <div
            class="chakra-stack css-0"
          >
            <h2
              class="chakra-heading css-0"
            >
              extra Shock content
            </h2>
            <p
              class="chakra-text css-0"
            >
              Shock content happened twice
            </p>
            <div
              class="css-0"
            >
              <p
                class="chakra-text css-0"
              >
                Author: Vasya
              </p>
              <p
                class="chakra-text css-0"
              >
                Thursday, September 29, 2022 at 10:00:28 AM GMT+2
              </p>
              <a
                class="chakra-link css-0"
                href="some url2"
                rel="noopener"
                target="_blank"
              >
                Read full article
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
